[Li Bailin] iP

rootkie opened the pull request:
# Duke Community Edition

> "Open source is a development methodology; free software is a social movement." ~ Richard Stallman ([source](https://www.azquotes.com/quote/755236?ref=open-source))

Duke (CE) is the **latest, free, feature rich** chatbot :rocket: based on the brilliant idea of Duke. It's,

* text-based
* easy to use
* FREE:fire:

## Installation

You can start using duke in a few simple steps

1. download it from [here](https://github.com/rootkie/ip/releases)
2. run it from command line. `java -jar ip.jar`
3. Start adding your tasks! 

## Features

- [x] Managing tasks
- [ ] Store tasks between usage
- [ ] Natural Language Processing
- [ ] Scheduling

## Contributors

If you are a java programmer, you are welcome to contribute to this *free* project. Here is the entry point of the project.

```java
import duke.Duke;

public class Main {

    public static void main(String[] args) {
        Duke duke = new Duke();
        duke.run();
    }
}
```

Commit:
Add Gradle support

rootkie committed:
Add functional reply and greet

rootkie committed:
Add echo and exit

rootkie committed:
Add task add and list

rootkie committed:
Change switch to if-else for level 3+

rootkie committed:
Add mark task as done

rootkie committed:
Add Deadline, Event, Todo class. Change Task to mutable

rootkie committed:
Add ToDo, Event, Deadline controls to Duke

Sidenote: removed return carriages, follow unix newline standard.

rootkie committed:
Fix scanner issue with automated testing

rootkie committed:
Add a test case to ui testing

rootkie committed:
Add exception for all possible errors

rootkie committed:
Update testcase for exception handling

rootkie committed:
Refactor complete task to align of general code structure

rootkie committed:
Add testcase for done argument exception handling

rootkie committed:
Add delete task and its testcases

rootkie committed:
Refactor id parsing and update relevant testcase

rootkie committed:
Make Duke main functional design for easy extension in future

rootkie committed:
Modify Task and subclasses to be more modular

rootkie committed:
Add save and load feature with testcases

rootkie committed:
Update gitignore to ignore db file

rootkie committed:
Change type of time for Deadline and Event to LocalDate with Testcases

rootkie committed:
Merge branch 'branch-Level-7'

rootkie committed:
Merge branch 'branch-Level-8'

rootkie committed:
Add more classes for better OOP

The classes are becoming too monolithic. Especially TaskManager and
Duke.

Extract functionality into their individual classes allow better
modularization and abstraction.

Let's,
* Abstract commands into their own classes
* Abstract File IO into Storage class
* Abstract User interaction into Ui class
* Make TaskManager non-static.

rootkie committed:
Merge branch 'A-MoreOOP'

rootkie committed:
Add packaging for duke, duke.command and duke.task

rootkie committed:
Merge branch 'packaging'

rootkie committed:
Remove unused factory methods in tasks

rootkie committed:
Change output date to a different format

rootkie committed:
Merge branch 'branch-Level-8'

rootkie committed:
Move all task processing from disk logic to Storage class

rootkie committed:
Add JUnit tests for multiple classes

rootkie committed:
Merge branch 'JUnit-Test'

rootkie committed:
Release v0.1 jar file

rootkie committed:
Fix bugs in tests

rootkie committed:
Add javadocs

rootkie committed:
Refactor CRLF to LF Unix standard

rootkie committed:
Refactor code style

rootkie committed:
Add find by keyword feature

rootkie committed:
Merge branch 'branch-A-CodingStandard'

rootkie committed:
Merge branch 'branch-Level-9'

rootkie committed:
Fix formatting issue from javadoc merge

rootkie committed:
Fix missing javadoc and import statements

rootkie committed:
Add checkstyle files from se-edu/addressbook-level3

rootkie committed:
Fix checkstyle errors

rootkie committed:
Merge branch 'branch-A-CheckStyle'

zognin commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,14 @@
+import duke.Duke;
+
+public class Main {
+
```
```suggestion
```
Perhaps you may want to remove unnecessary line breaks to make it consistent with your other classes?

zognin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    public static Command parse(String fullCommand) throws DukeException, IllegalArgumentException {
+        String[] parts = fullCommand.split(" ", 2);
+        String [] args;
```
```suggestion
        String[] args;

```
Is there a reason for putting a space between `String` and `[]`?

zognin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        case "deadline":
+            assertInputSize(parts, 2, "Not enough arguments");
+            args = parts[1].split(" /by ");
+            assertInputSize(args, 2, "☹ OOPS!!! Insufficient args for deadline.");
```
```suggestion
            assertDescriptionIsPresent(...);
            assertTimeIsPresent(...)
```
Perhaps you could follow the [SLAP](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#slap-hard) principle more closely?

rootkie committed:
Change encoding to UTF-8

rootkie committed:
Move checkstyle files to correct folder

rootkie committed:
Merge branch 'branch-A-CheckStyle'

zognin commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,102 @@
+     */
+    public Task deleteTask(int taskId) throws IllegalArgumentException {
+        if (taskId < taskList.size() && taskId >= 0) {
+            Task ret = taskList.remove(taskId);
```
Is there a reason why you named the variable `ret`? Is it a standard word because I can't think of what ret stands for?

zognin commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    public int taskCount() {
```
```suggestion
    public int getNumOfTasks() {
```
Perhaps you could name methods in verb form? For e.g. `getNumOfTasks`, `countTasks`, `getSize`

rootkie committed:
Fix checkstyle errors for tests

rootkie committed:
Merge branch 'branch-A-Gradle'

zognin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Deadline extends Task {
+
+    private final LocalDate time;
```
```suggestion
    private final LocalDate date;
```
Perhaps it would be better to name this date, as it is a `LocalDate` object and not a `LocalTime` object? This could explain what the variable is more accurately.

zognin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public String toDatabaseString() {
```
```suggestion
    public String toStorageString() {
```
Would it be clearer to use `Storage` instead of `Database`, since you have a `Storage` class to represent the stored data and not a `Database` class?

zognin commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,60 @@
+ * Represents a event type task.
+ */
+public class Event extends Task {
+
+
```
Similarly, you may want to remove unnecessary line breaks here to make it consistent with other classes

rootkie committed:
Add minimal working GUI feature

rootkie committed:
Change project structure and remove text-ui-test for GUI-based application

rootkie committed:
Merge branch 'branch-Level-10'

rootkie committed:
Fix UI non-responsive bug, interface now scales horizontally

rootkie committed:
Fix more UI issues

FergusMok commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+            return new DeleteCommand(Integer.parseInt(parts[1]) - 1);
+        case "find":
+            assertInputSize(parts, 2, "Not enough arguments");
+            return new FindCommand(parts[1]);
```
Would it be define the error messages as private static final variable instead? This would make your code adhere better to the [DRY Principle](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#minimize-code-duplication).

FergusMok commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Saves given list of tasks to file specified at instantiation.
+     * If file does not exist, it will be created.
+     * The list of tasks will be formatted in easily parsable format delimited
+     * by |.
+     *
+     * @param taskList The list of tasks to be saved.
+     * @throws IOException If something goes wrong with File operations.
+     */
+    public void save(List<Task> taskList) throws IOException {
+        boolean fileExists = Files.exists(path);
```
Would it better to rename fileExists to `hasFile` instead? This would adhere to the coding standard of using is, has, was at the start of boolean names.

FergusMok commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public List<Task> load() throws IOException, DateTimeException {
+        boolean fileExists = Files.exists(path);
+        List<Task> taskList = new ArrayList<>();
```
Would renaming this variable to `tasksList` be better? This would better adhere to the plural naming for collection variables.

FergusMok commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,102 @@
+            save();
+            return taskList.get(taskId);
+        } else {
+            throw new IllegalArgumentException("☹ OOPS!!! Task Index is invalid!!");
```
Would it be better the string as a variable, similar to the other mentioned one in Parser.java?

FergusMok reviewed (commented):
Good job on your code, your testing is very thorough and everything looks very well-made. Just a few small nits to adhere closer to the coding standards.

rootkie committed:
Add assertions to some parts of the code

rootkie committed:
Change Storage class to decouple from Task class

Storage class is coupled with Task class. This make it too specific to
the use case of saving and loading tasks which should be handled
elsewhere.

Decouple Storage class from Task class makes it more general.

As a step towards the decoupling, let's
* make Storage class save and load generic string as data to disk.
* make TaskManager class to handle raw string data to tasks.

rootkie committed:
Refactor code based on PR review nus-cs2103-AY2122S1/ip#115

PR review has identified some code quality improvements that can be made
to the code base.

Make the suggested changes to improve code quality.

Let's,
* Rename the neccessary methods and variables.
* Extract functions in parser to adhere to SLAP principle.

rootkie committed:
Merge pull request #1 from rootkie/branch-A-Assertions

Add assertions to some parts of the code

rootkie committed:
Merge branch 'master' into branch-A-CodeQuality

rootkie committed:
Merge pull request #2 from rootkie/branch-A-CodeQuality

Branch a code quality

rootkie committed:
Add github actions CI file

rootkie committed:
Fix issue found by CI

rootkie committed:
Add detect duplicate task feature

rootkie committed:
Merge branch 'branch-C-DetectDuplicates'

rootkie committed:
Add user guide
